rcljava/src/main/java/org/ros2/rcljava/time/TimeSource.java

```diff
@@ -0,0 +1,161 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ros2.rcljava.time;
+
+import org.ros2.rcljava.consumers.Consumer;
+import org.ros2.rcljava.node.Node;
+import org.ros2.rcljava.parameters.ParameterCallback;
+import org.ros2.rcljava.parameters.ParameterType;
+import org.ros2.rcljava.parameters.ParameterVariant;
+import org.ros2.rcljava.subscription.Subscription;
+import org.ros2.rcljava.time.Clock;
+import org.ros2.rcljava.time.ClockType;
+import org.ros2.rcljava.Time;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class TimeSource {
+  private Node node;
+  private boolean rosTimeIsActive;
+  private ArrayList<Clock> associatedClocks;
+  private Subscription<rosgraph_msgs.msg.Clock> clockSub;
+  private Time lastTimeSet;
+  private ParameterCallback simTimeCB;
+
+  public TimeSource() {
+    this(null);
+  }
+
+  public TimeSource(Node node) {
+    this.rosTimeIsActive = false;
+    this.associatedClocks = new ArrayList<Clock>();
+    this.clockSub = null;
+    this.lastTimeSet = new Time(0, 0, ClockType.ROS_TIME);
+    if (node != null) {
+      this.attachNode(node);
+    }
+  }
+
+  public boolean getRosTimeIsActive() {
+    return this.rosTimeIsActive;
+  }
+
+  class SubscriptionCallback implements Consumer<rosgraph_msgs.msg.Clock> {
+    private TimeSource timeSource;
+
+    SubscriptionCallback(TimeSource ts) {
+      this.timeSource = ts;
+    }
+
+    public void accept(final rosgraph_msgs.msg.Clock msg) {
+      Time timeFromMsg = new Time(msg.getClock(), ClockType.ROS_TIME);
+      this.timeSource.lastTimeSet = timeFromMsg;
+      for (Clock clock : this.timeSource.associatedClocks) {
+        clock.setRosTimeOverride(timeFromMsg);
+      }
+    }
+  }
+
+  public void setRosTimeIsActive(boolean enabled) {
+    if (this.rosTimeIsActive == enabled) {
+      return;
+    }
+    this.rosTimeIsActive = enabled;
+    for (Clock clock : this.associatedClocks) {
+      clock.setRosTimeIsActive(enabled);
+    }
+    if (enabled) {
+      if (this.node != null) {
+        this.clockSub = node.<rosgraph_msgs.msg.Clock>createSubscription(rosgraph_msgs.msg.Clock.class, "/clock", new SubscriptionCallback(this));
+      }
+    } else {
+      if (this.node != null && this.clockSub != null) {
+        this.node.removeSubscription(this.clockSub);
+        this.clockSub = null;
+      }
+    }
+  }
+
+  public void attachNode(Node node) {
+    if (this.node != null) {
+      detachNode();
+    }
```

Comment:
Just because that's how rclpy does it: https://github.com/ros2/rclpy/blob/master/rclpy/rclpy/time_source.py#L76 .  However, `rclcpp` does not do this, and I think it is a fairly unlikely situation.  I can go either way with it.

Reply:
ah, yes this is fine.
It sounds like this is missing in `rclcpp`.